core/src/main/java/hudson/FilePath.java

```diff
@@ -2655,6 +2668,7 @@ private void readFromTar(String name, File baseDir, InputStream in) throws IOExc
                         if (mode != 0 && !Functions.isWindows()) // be defensive
                             _chmod(f, mode);
                     }
+                    files.add(f);
```

Comment:
Rather than returning a list of files, I wonder if it would make more sense to notify a callback interface as files are processed (same in `tar` for `stash`). If you are running into trouble, such as slow performance, because of some particular files, it would be more helpful to see that in real time.

A bit tricky since the callback would in general need to be `SerializableOnlyOverRemoting`. You can safely transmit a `TaskListener` that way, though only if you `flush` at the end of the `Callable`.